### PR TITLE
feat(sql-udf): add hint display for non-existent functions

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xzhseh** [comment]
Current dependencies on/for this PR:
* **#15188** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15188?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `xzhseh/sql-udf-creation-better-hint-display`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15188?utm_source=stack-comment).

File: src/frontend/src/handler/create_sql_function.rs [hunk 1]
```diff
@@ -32,31 +32,56 @@ use crate::catalog::CatalogError;
 use crate::expr::{Expr, ExprImpl, Literal};
 use crate::{bind_data_type, Binder};
 
+/// The error type for hint display
+/// Currently we will try invalid item first
```
Comment (xzhseh) [inline src/frontend/src/handler/create_sql_function.rs:36]:
```suggestion
/// Currently we will try invalid parameter first
```

**Review by wangrunji0408 — APPROVED** [review]
LGTM.
Having a highlight of invalid tokens is definitely cool. But the implementation looks more and more hacky to me. I think we should start investigating on the span-based solutions. 🥵

File: src/frontend/src/handler/create_sql_function.rs [hunk 2]
```diff
@@ -32,31 +32,56 @@ use crate::catalog::CatalogError;
 use crate::expr::{Expr, ExprImpl, Literal};
 use crate::{bind_data_type, Binder};
 
+/// The error type for hint display
+/// Currently we will try invalid parameter first
+/// Then try to find non-existent functions
+enum ErrMsgType {
+    Parameter,
+    Function,
```
[the remark this answers is not in the document]
Comment (xzhseh) [reply src/frontend/src/handler/create_sql_function.rs:40]: Actually I don't think this two could merge 🤔
Without prior context on what the error message is for, we can't simply retrieve the hint target from the input message.

File: src/frontend/src/handler/create_sql_function.rs [hunk 3]
```diff
@@ -32,31 +32,56 @@ use crate::catalog::CatalogError;
...
-    ret
+/// Extract the target name to hint display
+/// according to the type of the error message item
+fn extract_hint_display_target(err_msg_type: ErrMsgType, invalid_msg: &str) -> Option<&str> {
+    match err_msg_type {
+        // e.g., [sql udf] failed to find named parameter <target name>
+        ErrMsgType::Parameter => invalid_msg.split_whitespace().last(),
+        // e.g., function <target name> does not exist
+        ErrMsgType::Function => {
+            let func = invalid_msg.split_whitespace().nth(1).unwrap_or("null");
```
[the remark this answers is not in the document]
Comment (xzhseh) [reply src/frontend/src/handler/create_sql_function.rs:78]: Error type can not indicate [function / invalid parameter] name I'm assuming, so it seems impossible? 🥵

**Pull request merged**